### ISSUE TITLE
Show a correct message to the user when upgrade incorrectly through WPK

### ISSUE THREAD

**Issue by mauromalara** [wazuh/wazuh#18436]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.5.1-rc1 | WPK Upgrade | Manager | Packages | Ubuntu 22.04 LTS |

## Description

While executing a [manual test in RC1](https://github.com/wazuh/wazuh/issues/18409) we found a lack of clarity in the error displayed when upgrading a macOS (**arm64**) agent through WPK using an **intel64** package.

### Steps to reproduce

1. Try to upgrade a macOS (arm64) agent using an intel64 WPK package (check the related issue for additional details)

### Expected results

- A clear message about what is going wrong (the user is making a mistake)

### Current results

- The following message is shown when upgrading:
```
Agent information not found in database
concurrent.futures.process._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/concurrent/futures/process.py", line 246, in _process_worker
    r = call_item.fn(*call_item.args, **call_item.kwargs)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.5.1-py3.9.egg/wazuh/core/cluster/dapi/dapi.py", line 247, in run_local
    data = f(**f_kwargs)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.5.1-py3.9.egg/wazuh/rbac/decorators.py", line 423, in wrapper
    result = func(*args, **kwargs) if not skip_execution else None
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.5.1-py3.9.egg/wazuh/agent.py", line 989, in upgrade_agents
    raise WazuhInternalError(error_code, cmd_error=True, extra_message=agent_result['message'])
wazuh.core.exception.WazuhInternalError: Error 1816 - Agent information not found in database
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.5.1-py3.9.egg/wazuh/core/cluster/dapi/dapi.py", line 292, in execute_local_request
    data = await asyncio.wait_for(task, timeout=timeout)
  File "/var/ossec/framework/python/lib/python3.9/asyncio/tasks.py", line 442, in wait_for
    return await fut
wazuh.core.exception.WazuhInternalError: Error 1816 - Agent information not found in database
Internal error:
```

**Comment by vikman90** [comment]
Let's make the error message more friendly for this version.

**Comment by mauromalara** [comment]
@GGP1 @vikman90 @Selutario

After running the same test for RC2 (#18503), we consider that the output is still not acceptable:

```bash
# /var/ossec/bin/agent_upgrade -a 001 -f /home/ubuntu/wazuh_agent_v4.5.1_macos_x86_64.wpk -F
The WPK package used does not fit with the agent's architecture
concurrent.futures.process._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/concurrent/futures/process.py", line 246, in _process_worker
    r = call_item.fn(*call_item.args, **call_item.kwargs)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.5.1-py3.9.egg/wazuh/core/cluster/dapi/dapi.py", line 247, in run_local
    data = f(**f_kwargs)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.5.1-py3.9.egg/wazuh/rbac/decorators.py", line 423, in wrapper
    result = func(*args, **kwargs) if not skip_execution else None
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.5.1-py3.9.egg/wazuh/agent.py", line 989, in upgrade_agents
    raise WazuhInternalError(error_code, cmd_error=True, extra_message=agent_result['message'])
wazuh.core.exception.WazuhInternalError: Error 1836 - The WPK package used does not fit with the agent's architecture
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.5.1-py3.9.egg/wazuh/core/cluster/dapi/dapi.py", line 292, in execute_local_request
    data = await asyncio.wait_for(task, timeout=timeout)
  File "/var/ossec/framework/python/lib/python3.9/asyncio/tasks.py", line 442, in wait_for
    return await fut
wazuh.core.exception.WazuhInternalError: Error 1836 - The WPK package used does not fit with the agent's architecture
Error 500: The WPK package used does not fit with the agent's architecture
```

Please, analyze the possible solutions and apply a new fix.

**Comment by GGP1** [comment]
## Update

Opened a pull request with changes to catch framework errors from the API side that were previously not taken into consideration. 

We tried showing the most user-friendly message possible, but we had to wrap the internal error received from the framework with another error in order to hide the traceback, raising the exception as it was catched didn't work for some reason that we couldn't identity (since both `WazuhInternalError` and `WazuhError` are catched in the script).

Right now, the message shown is the following:

```console
root@ip-172-31-88-29:/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.5.1-py3.9.egg/scripts# /var/ossec/framework/python/bin/python3 /var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.5.1-py3.9.egg/scripts/agent_upgrade.py -a 001 -f /home/ubuntu/wazuh_agent_v4.5.1_macos_x86_64.wpk -F
Error 400: Wazuh Internal Error: Error 1836 - The WPK package used does not fit with the agent's architecture
```

> Where `Wazuh Internal Error:` is what we tried to get rid of but we couldn't

**Comment by Selutario** [comment]
## Update
This problem has two variants. On the one hand, that the complete trace of the error is shown. On the other, that the error itself occurs.

### Error trace
Regarding the error trace, it should disappear after merging the following PR. The reason is that an unexpected error was obtained (not finding information in the DB of an agent that did exist in the `client.keys`):
- https://github.com/wazuh/wazuh/pull/18521

Unexpected errors are raised as a `WazuhInternalError`, and as a consequence, the trace is explicitly printed:
https://github.com/wazuh/wazuh/blob/18ad05746eadc817f4d2db0b35287d463b210ad1/framework/wazuh/core/cluster/dapi/dapi.py#L311-L314

Now, the raised error will be WazuhError, which is considered expected and therefore, no error trace is needed. 

### Agent information not found in database
The framework function for upgrading agents follows several steps in order. The ones we care about in this case are:
1. A list of registered agents is obtained:
  https://github.com/wazuh/wazuh/blob/18ad05746eadc817f4d2db0b35287d463b210ad1/framework/wazuh/agent.py#L916-L919
2. If any non-existent agent was included as a parameter, it is marked as a failed item and is excluded from the list to upgrade:
  https://github.com/wazuh/wazuh/blob/18ad05746eadc817f4d2db0b35287d463b210ad1/framework/wazuh/agent.py#L933-L955
3. An upgrade request is sent to the rest of the agents and the socket response for each one is analyzed.
  https://github.com/wazuh/wazuh/blob/18ad05746eadc817f4d2db0b35287d463b210ad1/framework/wazuh/agent.py#L961-L964

I have investigated if it could be due to a caching problem in the framework, since the function that gets the list of agents in step 1. is cached. However, this cache is created right at the start of the upgrade request and is invalidated at the end, so this doesn't seem to be the problem.

It could be due to a race condition. That is, the agent existed when the framework got the list of active agents, but it was removed just before the upgrade request was sent to the socket. It seems very unlikely.

It would be necessary to reproduce the use case of the test to see if it is a consistent error and with well-defined causes or if it has been something fortuitous.

**Comment by vikman90** [comment]
The last reported problem will be fixed at #18547.

Closing this issue.
